File: plugins/plugin-bootstrap/python/elizaos_plugin_bootstrap/evaluators/goal.py
"""
Goal Evaluator - Evaluates progress toward goals and objectives.

This evaluator assesses how well the agent is progressing toward
defined goals and objectives.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Evaluator, ModelType

from elizaos_plugin_bootstrap.types import EvaluatorResult
from elizaos_plugin_bootstrap.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


GOAL_EVALUATION_TEMPLATE = """# Task: Evaluate progress toward goals.

{{providers}}

# Current Goals:
{{goals}}

# Recent Actions:
{{recentActions}}

# Instructions:
Evaluate how well the recent actions are contributing to the defined goals.
Rate progress on a scale of 0-100 and provide specific feedback.

Respond using XML format like this:
<response>
    <thought>Your analysis of goal progress</thought>
    <progress>Numeric progress score 0-100</progress>
    <feedback>Specific feedback on what's working and what isn't</feedback>
    <next_steps>Suggested next steps to make progress</next_steps>
</response>

IMPORTANT: Your response must ONLY contain the <response></response> XML block above."""


async def evaluate_goal_progress(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> EvaluatorResult:
    """
    Evaluate progress toward goals.

    Returns:
    - Progress score (0-100)
    - Specific feedback
    - Suggested next steps
    """
    if state is None:
        return EvaluatorResult(
            score=0,
            passed=False,
            reason="No state provided for evaluation",
            details={},
        )

    try:
        # Get goals from state or world
        goals: list[str] = []
        if state.values:
            goals_data = state.values.get("goals", [])
            if isinstance(goals_data, list):
                goals = [str(g) for g in goals_data]

        if not goals:
            return EvaluatorResult(
                score=100,
                passed=True,
                reason="No goals defined",
                details={"noGoals": True},
            )

        # Get recent actions from state
        recent_actions: list[str] = []
        if state.values:
            actions_data = state.values.get("recentActions", [])
            if isinstance(actions_data, list):
                recent_actions = [str(a) for a in actions_data]

        # Format goals and actions for prompt
        goals_text = "\n".join(f"- {goal}" for goal in goals)
        actions_text = (
            "\n".join(f"- {action}" for action in recent_actions)
            if recent_actions
            else "No recent actions"
        )

        prompt = runtime.compose_prompt(
            state=state,
            template=runtime.character.templates.get(
                "goalEvaluationTemplate", GOAL_EVALUATION_TEMPLATE
            ),
        )
        prompt = prompt.replace("{{goals}}", goals_text)
        prompt = prompt.replace("{{recentActions}}", actions_text)

        response_text = await runtime.use_model(ModelType.TEXT_LARGE, prompt=prompt)
        parsed_xml = parse_key_value_xml(response_text)

        if parsed_xml is None:
            raise ValueError("Failed to parse evaluation response")

        progress_str = str(parsed_xml.get("progress", "0"))
        try:
            progress = max(0, min(100, int(progress_str)))
        except ValueError:
            progress = 0

        thought = str(parsed_xml.get("thought", ""))
        feedback = str(parsed_xml.get("feedback", ""))
        next_steps = str(parsed_xml.get("next_steps", ""))

        # Determine if evaluation passed (progress >= 50)
        passed = progress >= 50

        return EvaluatorResult(
            score=progress,
            passed=passed,
            reason=feedback,
            details={
                "thought": thought,
                "feedback": feedback,
                "nextSteps": next_steps,
                "goalCount": len(goals),
                "actionCount": len(recent_actions),
            },
        )

    except Exception as error:
        runtime.logger.error(
            {
                "src": "evaluator:goal",
                "agentId": runtime.agent_id,
                "error": str(error),
            },
            "Error evaluating goal progress",
        )
        return EvaluatorResult(
            score=0,
            passed=False,
            reason=f"Evaluation error: {str(error)}",
            details={"error": str(error)},
        )


async def validate_goal_evaluation(
    runtime: IAgentRuntime,
    message: Memory,
) -> bool:
    """Validate that goal evaluation can be performed."""
    # Goals can always be evaluated
    return True


# Create the evaluator instance
goal_evaluator = Evaluator(
    name="GOAL",
    description="Evaluates progress toward defined goals and objectives",
    validate=validate_goal_evaluation,
    handler=evaluate_goal_progress,
    examples=[],  # Examples can be added later
)

File: plugins/plugin-bootstrap/rust/src/evaluators/goal.rs
//! GOAL evaluator implementation.

use async_trait::async_trait;

use crate::error::{PluginError, PluginResult};
use crate::runtime::{IAgentRuntime, ModelParams};
use crate::types::{EvaluatorResult, Memory, ModelType, State};
use crate::xml::parse_key_value_xml;

use super::Evaluator;

const GOAL_EVALUATION_TEMPLATE: &str = r#"# Task: Evaluate progress toward goals.

{{providers}}

# Current Goals:
{{goals}}

# Recent Actions:
{{recentActions}}

# Instructions:
Evaluate how well the recent actions are contributing to the defined goals.
Rate progress on a scale of 0-100 and provide specific feedback.

Respond using XML format like this:
<response>
    <thought>Your analysis of goal progress</thought>
    <progress>Numeric progress score 0-100</progress>
    <feedback>Specific feedback on what's working and what isn't</feedback>
    <next_steps>Suggested next steps to make progress</next_steps>
</response>

IMPORTANT: Your response must ONLY contain the <response></response> XML block above."#;

/// Evaluator for goal progress.
pub struct GoalEvaluator;

#[async_trait]
impl Evaluator for GoalEvaluator {
    fn name(&self) -> &'static str {
        "GOAL"
    }

    fn description(&self) -> &'static str {
        "Evaluates progress toward defined goals and objectives"
    }

    async fn validate(&self, _runtime: &dyn IAgentRuntime, _message: &Memory) -> bool {
        true
    }

    async fn evaluate(
        &self,
        runtime: &dyn IAgentRuntime,
        message: &Memory,
        state: Option<&State>,
    ) -> PluginResult<EvaluatorResult> {
        let state = match state {
            Some(s) => s,
            None => {
                return Ok(EvaluatorResult::fail(0, "No state provided for evaluation"));
            }
        };

        // Get goals from state
        let goals: Vec<String> = state
            .values
            .get("goals")
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|v| v.as_str().map(String::from))
                    .collect()
            })
            .unwrap_or_default();

        if goals.is_empty() {
            return Ok(EvaluatorResult::pass(100, "No goals defined")
                .with_detail("noGoals", true));
        }

        // Get recent actions from state
        let recent_actions: Vec<String> = state
            .values
            .get("recentActions")
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|v| v.as_str().map(String::from))
                    .collect()
            })
            .unwrap_or_default();

        // Format for prompt
        let goals_text = goals
            .iter()
            .map(|g| format!("- {}", g))
            .collect::<Vec<_>>()
            .join("\n");

        let actions_text = if recent_actions.is_empty() {
            "No recent actions".to_string()
        } else {
            recent_actions
                .iter()
                .map(|a| format!("- {}", a))
                .collect::<Vec<_>>()
                .join("\n")
        };

        // Get template and compose prompt
        let template = runtime
            .character()
            .templates
            .get("goalEvaluationTemplate")
            .map(|s| s.as_str())
            .unwrap_or(GOAL_EVALUATION_TEMPLATE);

        let composed_state = runtime
            .compose_state(message, &["RECENT_MESSAGES"])
            .await?;

        let prompt = runtime
            .compose_prompt(&composed_state, template)
            .replace("{{goals}}", &goals_text)
            .replace("{{recentActions}}", &actions_text);

        // Call the model
        let response = runtime
            .use_model(ModelType::TextLarge, ModelParams::with_prompt(&prompt))
            .await
            .map_err(|e| PluginError::ModelError(e.to_string()))?;

        let response_text = response
            .as_text()
            .ok_or_else(|| PluginError::ModelError("Expected text response".to_string()))?;

        // Parse XML response
        let parsed = parse_key_value_xml(response_text)
            .ok_or_else(|| PluginError::XmlParse("Failed to parse evaluation response".to_string()))?;

        let thought = parsed.get("thought").cloned().unwrap_or_default();
        let progress_str = parsed.get("progress").cloned().unwrap_or_else(|| "0".to_string());
        let feedback = parsed.get("feedback").cloned().unwrap_or_default();
        let next_steps = parsed.get("next_steps").cloned().unwrap_or_default();

        let progress: u8 = progress_str
            .parse()
            .unwrap_or(0)
            .min(100);

        let passed = progress >= 50;

        let result = if passed {
            EvaluatorResult::pass(progress, &feedback)
        } else {
            EvaluatorResult::fail(progress, &feedback)
        };

        Ok(result
            .with_detail("thought", thought)
            .with_detail("feedback", feedback)
            .with_detail("nextSteps", next_steps)
            .with_detail("goalCount", goals.len() as i64)
            .with_detail("actionCount", recent_actions.len() as i64))
    }
}

